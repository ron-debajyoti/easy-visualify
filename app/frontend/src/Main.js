import React, {Component} from 'react'
import App from './App'
import PlayWidget from './Widget'
import {Link} from 'react-router-dom'
import ReactTooltip from 'react-tooltip'
import styled from 'styled-components/macro'
import UserModal from './UserModal'
import * as util from './utils/Utility'
import Cookies from 'js-cookie'
var _ = require('lodash')

const Button = styled.button`
  background: palevioletred;
  color: white;
  font-size: 0.75em;
  margin: 1em;
  padding: 0.5em 1em;
  border: 2px solid black;
  border-radius: 3px;
`;

const TriggerButton = styled(Button)`
  float: right;
`


const ButtonWrapper = styled.div`
  display: inline-block;
  background: white;
`

const Wrapper = styled.div`
  background-color: white; 
`



const WidgetTitle = styled.div`
  font-size: 1.5em;
  float: center;
  text-align: left;
  color: palevioletred;
  background-color: white;
`

const WidgetWrapper = styled.div`
  padding: 1em;
  background : aquamarine
  float: right;
`




class Main extends Component{
    state = {
      content : "s",
      country : "None",
      contentType : 'top',
      user : {
        display_name: null,
        country: null,
        followers: null,
        images:null,
        external_urls: null,
        product: null,
        topTracks: null,
        topArtists: null,
        recommendedGenre: null
      },
      allUpdated :false
    }
  
    componentDidMount(){
      // let accessToken = queryString.parse(window.location.search)['?access_token']
      // console.log(queryString.parse(window.location.search))
      // console.log(accessToken)
      var access_token = Cookies.get('access_token')
      console.log(access_token)

      Promise.all(
        [util.fetchUserData(access_token)
        .then(data => {
          this.setState(() => ({
            ...this.state,
            user : {
              ...this.state.user,
              display_name: data.display_name,
              country: data.country,
              followers: data.followers.total,
              external_urls: data.external_urls,
              images: data.images,
              product:data.product
            }
          }))
        }),
        util.fetchTopTracks(access_token)
          .then(data => {
            this.setState(() => ({
              ...this.state,
              user : {
                ...this.state.user,
                topTracks : data.items
              }
            }))
          }),        
        util.fetchTopArtists(access_token)
          .then(data => {
            this.setState(() => ({
              ...this.state,
              user : {
                ...this.state.user,
                topArtists : data
              }
            }))
          })
          .then(() => this.updateGenre())
          .then(genre => {
            this.setState(() => ({
              ...this.state,
              user : {
                ...this.state.user,
                recommendedGenre : genre
              }
            }))
          })
        ])
      .then(() => {
        this.setState(() => ({
          ...this.state,
          allUpdated : true
        }))
      })
      .then(() => console.log(this.state))
      
      
    }
  
  
    updateGenre = async () => {
        var genre = []
        this.state.user.topArtists.items.slice(0,10).map((artist)=> {
          var temp = artist.genres
          genre = genre.concat(temp)
          genre = _.uniq(genre)
        })
        return genre
    }



    onUpdate = (content) => {
      this.setState(() => ({
        content
        })
      )
      console.log("updated state !")
      console.log(this.state.contentType)
      console.log(this.state.content)
    }
  
  
    onButtonClick = (type) => {
        this.setState(() => ({
          contentType : type
        }))
    }

    tooltipRender = (content) => {
        // console.log("!!!!")
        // console.log(content)
        this.setState(() => ({
          country:content
        }))
    }
  
  
    IsValidData = () => {
      if (this.state.content.length >=2 ){
        if(this.state.contentType === "top"){
          return (
            <WidgetWrapper>
              <WidgetTitle> Top 10 Playlists of {this.state.content[2]} </WidgetTitle> 
              <ol> 
                <PlayWidget width={300} height={100} uri={this.state.content[0][0].uri} />
                <PlayWidget width={300} height={100} uri={this.state.content[0][1].uri} />
                <PlayWidget width={300} height={100} uri={this.state.content[0][2].uri} />
                <PlayWidget width={300} height={100} uri={this.state.content[0][3].uri} />
                <PlayWidget width={300} height={100} uri={this.state.content[0][4].uri} />
                <PlayWidget width={300} height={100} uri={this.state.content[0][5].uri} />
                <PlayWidget width={300} height={100} uri={this.state.content[0][6].uri} />
                <PlayWidget width={300} height={100} uri={this.state.content[0][7].uri} />
                <PlayWidget width={300} height={100} uri={this.state.content[0][8].uri} />
                <PlayWidget width={300} height={100} uri={this.state.content[0][9].uri} />
              </ol>
            </WidgetWrapper>
          )
        }
        else{
          return (
            <WidgetWrapper>
              <WidgetTitle> Viral 10 Playlists of {this.state.content[2]} </WidgetTitle>
              <ol >
                <PlayWidget width={300} height={100} uri={this.state.content[1][0].uri} />
                <PlayWidget width={300} height={100} uri={this.state.content[1][1].uri} />
                <PlayWidget width={300} height={100} uri={this.state.content[1][2].uri} />
                <PlayWidget width={300} height={100} uri={this.state.content[1][3].uri} />
                <PlayWidget width={300} height={100} uri={this.state.content[1][4].uri} />
                <PlayWidget width={300} height={100} uri={this.state.content[1][5].uri} />
                <PlayWidget width={300} height={100} uri={this.state.content[1][6].uri} />
                <PlayWidget width={300} height={100} uri={this.state.content[1][7].uri} />
                <PlayWidget width={300} height={100} uri={this.state.content[1][8].uri} />
                <PlayWidget width={300} height={100} uri={this.state.content[1][9].uri} />
              </ol>
            </WidgetWrapper>
          )
        }
      }
      else{
        return (<WidgetTitle> Spotify is not supported in {this.state.content[0]} !</WidgetTitle>)
      }
    }
  
    render() {
      if(this.state.allUpdated){
        return(
          <Wrapper> 
              {/* {console.log(this.state.user)} */}
              <h1 style={{'textAlign' :'center'}}>Welcome</h1>
              <UserModal userObject={this.state.user} />
              <Link to='/main'>
                <App setTooltipContent={(e) => this.onUpdate(e)} setTooltip={this.tooltipRender}/>
              </Link>
              <ButtonWrapper>
                <TriggerButton className="Top10" onClick={() => this.onButtonClick("top")}> View Top 10 </TriggerButton>
                <TriggerButton className="Viral10" onClick={() => this.onButtonClick("viral")}> View Viral 10 </TriggerButton>
              </ButtonWrapper>
              <this.IsValidData />
              <ReactTooltip>{this.state.country}</ReactTooltip>
          </Wrapper>  
        );
      }
      else{
        console.log('okay somethings wrong here')
        return(
          <h1>Please wait!</h1>
        )
      }
      
    }


  
    // end of Main Class
  }


  export default Main;
  