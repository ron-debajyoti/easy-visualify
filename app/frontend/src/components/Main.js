import React, {Component} from 'react'
import App from './App'
import PlayWidget from './Widget'
import {Link} from 'react-router-dom'
import ReactTooltip from 'react-tooltip'
import styled from 'styled-components/macro'
import UserModal from './UserModal'
import * as util from '../utils/Utility'
import Cookies from 'js-cookie'
var _ = require('lodash')

const Button = styled.button`
  background: palevioletred;
  color: white;
  font-size: 0.75em;
  margin: 1em;
  padding: 0.5em 1em;
  border: 2px solid black;
  border-radius: 3px;
`;

const TriggerButton = styled(Button)`
  float: right;
`


const ButtonWrapper = styled.div`
  display: inline-block;
  background: white;
`

const Wrapper = styled.div`
  background-color: white; 
`



const WidgetTitle = styled.div`
  font-size: 1.5em;
  float: center;
  margin: 50px 10px
  text-align: left;
  color: palevioletred;
  background-color: white;
`

const WidgetWrapper = styled.div`
  padding: 1em;
  background : aquamarine
  float: right;
`




class Main extends Component{
    state = {
      content : "s",
      country : "None",
      contentType : 'top',
      user : {
        display_name: null,
        country: null,
        followers: null,
        images:null,
        external_urls: null,
        product: null,
        topTracks: null,
        topArtists: null,
        recommendedGenre: null,
        userPlaylists: null,
      },
      allUpdated :false
    }
  
    componentDidMount(){
      // let accessToken = queryString.parse(window.location.search)['?access_token']
      // console.log(queryString.parse(window.location.search))
      // console.log(accessToken)
      var access_token = Cookies.get('access_token')
      Promise.all(
        [util.fetchUserData(access_token)
        .then(data => {
          this.setState(() => ({
            ...this.state,
            user : {
              ...this.state.user,
              display_name: data.display_name,
              country: data.country,
              followers: data.followers.total,
              external_urls: data.external_urls,
              images: data.images,
              product:data.product
            }
          }))
        }),
        util.fetchTopTracks(access_token)
          .then(data => {
            this.setState(() => ({
              ...this.state,
              user : {
                ...this.state.user,
                topTracks : data.items
              }
            }))
          }),     
        util.fetchUserPlaylists(access_token)
          .then(data => {
            this.setState(() => ({
              ...this.state,
              user : {
                ...this.state.user,
                userPlaylists: data.items
              }
            }))
          }),
        util.fetchTopArtists(access_token)
          .then(data => {
            this.setState(() => ({
              ...this.state,
              user : {
                ...this.state.user,
                topArtists : data
              }
            }))
          })
          .then(() => this.updateGenre())
          .then(genre => {
            this.setState(() => ({
              ...this.state,
              user : {
                ...this.state.user,
                recommendedGenre : genre
              }
            }))
          })
        ])
      .then(() => {
        this.setState(() => ({
          ...this.state,
          allUpdated : true
        }))
      })
      
    }
  
  
    updateGenre = async () => {
        var genre = []
        this.state.user.topArtists.items.slice(0,15).map(artist => {
          var temp = artist.genres
          genre = genre.concat(temp)
          genre = _.uniq(genre)
        })
        return genre
    }



    onUpdate = (content) => {
      this.setState(() => ({
        content
        })
      )
      console.log("updated state !")
      // console.log(this.state.contentType)
      // console.log(this.state.content)
    }
  
  
    onButtonClick = (type) => {
        this.setState(() => ({
          contentType : type
        }))
    }

    tooltipRender = (content) => {
        // console.log("!!!!")
        // console.log(content)
        this.setState(() => ({
          country:content
        }))
    }
  
  
    IsValidData = () => {
      if (this.state.content.length >=2 ){
        var populateCards1 = this.state.content[0].map( element => <PlayWidget key={element.uri} width={300} height={100} uri={element.uri} />)
        var populateCards2 = this.state.content[1].map( element => <PlayWidget key={element.uri} width={300} height={100} uri={element.uri} />)
        var populateCards3 = this.state.content[2].map( element => <PlayWidget key={element.uri} width={300} height={100} uri={element.uri} />)

        if(this.state.contentType === "top"){
          if(this.state.content[0].length > 0){
            return (
              <WidgetWrapper>
                <WidgetTitle> Top 10 Tracks of {this.state.content[3]} </WidgetTitle>
                
                <ol> 
                 {populateCards1}
                </ol>
              </WidgetWrapper>
            )
          } else {
            return (
            <WidgetTitle>
              <WidgetTitle style={{'fontSize':'1em'}}> Spotify doesn't have the data yet.
               Try Radar Tracks   </WidgetTitle>
            </WidgetTitle>
            )
          }
        } else {
          if(this.state.contentType === "viral"){
            if (this.state.content[1].length > 0){
              return (
                <WidgetWrapper>
                  <WidgetTitle> Viral 10 Tracks of {this.state.content[3]} </WidgetTitle>
                  <ol >
                    {populateCards2}
                  </ol>
                </WidgetWrapper>
              )
            } else {
              return (
                <WidgetTitle>
                  <WidgetTitle style={{'fontSize':'1em'}}> Spotify doesn't have the data yet.
                   Try Radar Tracks </WidgetTitle>
                </WidgetTitle>
                )
            }
          } else {
            if (this.state.content[2].length > 0){
              return (
                <WidgetWrapper>
                  <WidgetTitle> Radar Tracks of {this.state.content[3]} </WidgetTitle>
                  <ol >
                    {populateCards3}
                  </ol>
                </WidgetWrapper>
              )
            } else {
              return (
                <WidgetTitle>
                  <WidgetTitle style={{'fontSize':'1em'}}> Spotify doesn't have the data yet.
                   Try other Tracks </WidgetTitle>
                </WidgetTitle>
                )
            }

          }
        }
      }
      else{
        if(this.state.content[0] === 's'){
          return (<WidgetTitle> Click on a country to begin! </WidgetTitle>)
        }
        return (<WidgetTitle> Spotify is not supported in {this.state.content[0]} !</WidgetTitle>)
      }
    }
  
    render() {
      if(this.state.allUpdated){
        return(
          <Wrapper> 
              {/* {console.log(this.state.user)} */}
              <h1 style={{'textAlign' :'center'}}>Welcome</h1>
              <UserModal userObject={this.state.user} />
              <Link to='/main'>
                <App setTooltipContent={(e) => this.onUpdate(e)} setTooltip={this.tooltipRender}/>
              </Link>
              <ButtonWrapper>
                <TriggerButton className="Top10" onClick={() => this.onButtonClick("top")}> View Top 10 Tracks</TriggerButton>
                <TriggerButton className="Viral10" onClick={() => this.onButtonClick("viral")}> View Viral 10 Tracks </TriggerButton>
                <TriggerButton className="Radar" onClick={() => this.onButtonClick("radar")}> View Radar Tracks </TriggerButton>
              </ButtonWrapper>
              <this.IsValidData />
              <ReactTooltip>{this.state.country}</ReactTooltip>
          </Wrapper>  
        );
      }
      else{
        console.log('okay page is loading')
        return(
          <h3 style={{'textAlign':'center'}}>Loading ... </h3>
        )
      }
      
    }


  
    // end of Main Class
  }


  export default Main;
  