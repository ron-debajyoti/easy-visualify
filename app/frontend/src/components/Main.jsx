import React, { useState, useEffect } from 'react';
import ReactTooltip from 'react-tooltip';
import styled from 'styled-components/macro';
import media from 'styled-media-query';
import Cookies from 'js-cookie';
import App from './App';
import UserModal from './UserModal';
import SongRenderer from './SongRenderer';
import * as util from '../utils/Utility';

const _ = require('lodash');

const Button = styled.button`
  background: palevioletred;
  color: white;
  ${media.lessThan('medium')`
    font-size: 0.75vm;
    margin: 2vm;
    padding: 0.5vm 1vm;
  `}
  ${media.greaterThan('medium')`
    font-size: 0.75em;
    margin: 1em;
    padding: 0.5em 1em;
  `}
  border: 2px solid black;
  border-radius: 3px;
`;

const TriggerButton = styled(Button)`
  float: right;
`;

const ButtonWrapper = styled.div`
  display: inline-grid;
`;

const Wrapper = styled.main``;

const Section1 = styled.div`
  width: 70%;
  height: 100%;
  float: left;
`;

const Section2 = styled.div`
  display: inline-grid;
  height: 100%;
  width: 30%;
`;

const MainTitle = styled.h2`
  ${media.lessThan('medium')`
    font-size: large;
  `}

  color: white;
  text-align: start;
  background: black;
`;

const Main = () => {
  /* setting various hooks here  */
  const [content, setContent] = useState([]);
  const [country, setCountry] = useState('None');
  const [contentType, setContentType] = useState('top');
  const [user, setUser] = useState({
    user: {
      display_name: null,
      country: null,
      followers: null,
      images: null,
      external_urls: null,
      product: null,
      topTracks: null,
      topArtists: null,
      recommendedGenre: null,
      userPlaylists: null,
    },
  });
  const [allUpdated, setAllUpdated] = useState(false);

  const updateGenre = async (data) => {
    let genre = [];
    data.topArtists.items.slice(0, 15).map((artist) => {
      const temp = artist.genres;
      genre = genre.concat(temp);
      genre = _.uniq(genre);
      return true;
    });
    return genre;
  };

  useEffect(() => {
    // let accessToken = queryString.parse(window.location.search)['?access_token']
    // console.log(queryString.parse(window.location.search))
    // console.log(accessToken)
    const accessToken = Cookies.get('access_token');
    let userData = {
      display_name: null,
      country: null,
      followers: null,
      images: null,
      external_urls: null,
      product: null,
      topTracks: null,
      topArtists: null,
      recommendedGenre: null,
      userPlaylists: null,
    };
    Promise.all([
      util.fetchUserData(accessToken).then((data) => {
        userData = {
          ...userData,
          display_name: data.display_name,
          country: data.country,
          followers: data.followers.total,
          external_urls: data.external_urls,
          images: data.images,
          product: data.product,
        };
      }),
      util.fetchTopTracks(accessToken).then((data) => {
        userData = {
          ...userData,
          topTracks: data.items,
        };
      }),
      util.fetchUserPlaylists(accessToken).then((data) => {
        userData = {
          ...userData,
          userPlaylists: data.items,
        };
      }),
      util
        .fetchTopArtists(accessToken)
        .then((data) => {
          userData = {
            ...userData,
            topArtists: data,
          };
        })
        .then(() => updateGenre(userData))
        .then((genre) => {
          userData = {
            ...userData,
            recommendedGenre: genre,
          };
        }),
    ]).then(() => {
      console.log('!!!!!!!!!!11');
      setUser(userData);
      setAllUpdated(true);
    });
  }, []);

  const onUpdate = (receivedContent) => {
    setContent(receivedContent);
    console.log('updated state !');
  };

  const onButtonClick = (type) => {
    setContentType(type);
  };

  const tooltipRender = (receivedcountry) => {
    setCountry(receivedcountry);
    ReactTooltip.rebuild();
  };

  // useEffect(() => {

  // }, [contentType])

  return (
    <div>
      {allUpdated ? (
        <Wrapper>
          <MainTitle> Main Page </MainTitle>
          <Section1>
            <UserModal userObject={user} />
            <App setTooltipContent={(e) => onUpdate(e)} setTooltip={tooltipRender} />
          </Section1>
          <Section2>
            <ButtonWrapper>
              <TriggerButton className="Top10" onClick={() => onButtonClick('top')}>
                {' '}
                View Top 10 Tracks
              </TriggerButton>
              <TriggerButton className="Viral10" onClick={() => onButtonClick('viral')}>
                {' '}
                View Viral 10 Tracks{' '}
              </TriggerButton>
              <TriggerButton className="Radar" onClick={() => onButtonClick('radar')}>
                {' '}
                View Radar Tracks{' '}
              </TriggerButton>
            </ButtonWrapper>
            <SongRenderer content={content} contentType={contentType} />
          </Section2>
          <ReactTooltip>{country}</ReactTooltip>
        </Wrapper>
      ) : (
        <h3 style={{ textAlign: 'center', color: 'white' }}>Loading ... </h3>
      )}
    </div>
  );
};

export default Main;
