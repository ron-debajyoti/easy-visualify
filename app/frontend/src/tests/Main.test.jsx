// jest.mock('');
import React from 'react';
import { unmountComponentAtNode } from 'react-dom';
import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import renderer from 'react-test-renderer';
import Main from '../components/Main';

let container = null;

jest.mock('../utils/Utility');

const waitForComponent = async (wrapper) => {
  await act(async () => {
    await new Promise((resolve) => setTimeout(resolve, 0));
    wrapper.update();
    console.log('!!!');
  });
};

// const mockUseEffect = () => {
//   useEffect.mockImplementationOnce((f) => f());
// };

beforeEach(() => {
  // setup a DOM element as a render target
  container = document.createElement('div');
  document.body.appendChild(container);
});

afterEach(() => {
  // cleanup on exitingq
  jest.clearAllMocks();
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

/* Tests */
describe('Testing the Main component : ', () => {
  it('check loading snapshot for Main', () => {
    const tree = renderer.create(<Main />, container).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should fetch user and playlist data', () => {
    const wrapper = mount(<Main />, container);
    waitForComponent(wrapper);
    expect(wrapper.exists('section-2')).toBeTruthy();
    expect(wrapper.find('section-1').length).toBe(1);
    expect(wrapper.find('h2')).toEqual('Main Page');
  });
});
