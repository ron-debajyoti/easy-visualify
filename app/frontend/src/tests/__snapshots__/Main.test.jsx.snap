// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing the Main component :  check loading snapshot for Main 1`] = `
.c1 {
  color: white;
  font-size: 1.5em;
  text-align: center;
}

.c0 {
  text-align: center;
}

.c2 {
  max-width: 100%;
  margin: 10px;
  object-fit: contain;
}

<div>
  <div
    className="c0"
  >
    <div
      className="c1"
    >
       Authenticating 
    </div>
    <img
      alt="loading..."
      className="c2"
      src="auth.gif"
    />
  </div>
</div>
`;
